src/socketio.quiver.js

```diff
@@ -263,23 +300,59 @@ QuiverSocialProvider.prototype.login = function(loginOpts, continuation) {
             this.connectAsClient_.bind(this, friend, null)).
                 catch(onConnectionFailure.bind(this, friend)));
       }
-      Promise.all(connectionPromises).then(finishLogin);
+      return Promise.all(connectionPromises);
     }.bind(this);
 
     this.connectLoop_(this.configuration_.self.servers,
-        this.connectAsOwner_.bind(this)).then(connectToFriends, connectToFriends);
+        this.connectAsOwner_.bind(this)).then(function(results) {
+      if (results.succeeded.length > 0) {
+        // If at least one connection succeeded, then we do have a working
+        // network, so any connection failures are because the server is in fact
+        // failing or unreachable.  Remove it from the active list.
+        // It will then be replaced by a friend's server.
+        results.failed.forEach(this.removeServer_, this);
+      }
+      return connectToFriends();
+    }.bind(this)).then(function() {
+      var deficit = QuiverSocialProvider.MAX_CONNECTIONS_ - this.countOwnerConnections_();
+      if (deficit > 0) {
+        /** @type {!Object<string, QuiverSocialProvider.server_>} */
+        var unusedDefaultServers = {};
+            QuiverSocialProvider.DEFAULT_SERVERS_.forEach(function(server) {
```

Comment:
This code is for the fallback case where one of our initial servers has failed, and we have not managed to contact any friends who are using any other server, so even after contacting all our friends we are still low on servers.

>  for each one you haven't yet connected to as a client, you connect to them as an owner.

Yes.  However, the process of connecting to a server as a client also checks if we have too few owner-servers, and if so, connects as an owner as well.  Therefore, we should only get to this point if `this.connections_` contains less than DEFAULT_SERVERS (i.e. 2) connections.

> Shouldn't you instead be connecting to the this.configuration_.self.servers instead?

No, we've already done that, and removed servers that failed.  This is to fill in any missing servers.

> With the current code is it possible that unusedDefaultServers will end up being empty (if you've connected as a client to all of your DEFAULT_SERVERS)?

Currently, `unusedDefaultServers` will almost always be empty because there's only one default server!  However, if there are more default servers, then `unusedDefaultServers` should never be empty (when `deficit > 0`).